### ISSUE TITLE
hydratation and rerendering

### ISSUE THREAD

**Issue by rodbs** [hashicorp/next-mdx-remote#84]
I have an array of content that I hydrate separately, so eventually I gets as many re-renders as elements in the array.
I'd like to avoid all these re-renders while keeping the content separated in the array  because I need to control it on displaying it.

Any ideas? merging the content before to hydrate it just once? but how I keep track of each section later?  
Thanks!

**Comment by jescalan** [comment]
What do you need to control later? Can this be accomplished via an embedded component? Super interesting question, just need a little more context 😀

**Comment by rodbs** [comment]
I'm processing/hydrating the content by section  (in a array) because I want to display it by sections (with a menu), instead of in a whole page. 
Also, on displaying I'm activating each section with display:none/block because I want to make all content available in the DOM for the google index bot.
Now, when switching sections, the main page gets re-rerender and every section gets  rehydrated and re-render.  So I'm getting as many re-render as sections has the page.
https://inglesando.com/curso-ingles-online/preguntas-ingles/preguntas-basicas-en-ingles-this-it-what

Basically, I need to have all content on the DOM but managing it by section. And ideally with one render on hydration. Might I use useMemo on hydrated section content? ...
What do you mean by embedded component?

**Comment by jescalan** [comment]
Got it - there's no way for you to run these as separate files? This would make it much easier

**Comment by rodbs** [comment]
What do you mean as "separated files"?  The problem I'm facing now is that I'm hydrating it separated. I'd like to hydrate sections together and them display them separated. I'd like to have one hydration per page. 
I guess my question/issue is if there is a way on hydrating to leave some kind of marks that I can read them on displaying to control it. (html ids ....)

**Comment by jescalan** [comment]
The standard use case for this library is to read and display a single markdown document per page. With a classic blog example, you'd get a blog post and display it on each post page. You could do it like this here by having your different sections each render as their own pages, but it seems like you have the pushed into a single page. Because of this, the back button is worthless as you go through sections, and there's no way to link to a section within a lesson.

I guess what I'm wondering is whether you could shift this back to a more standard architecture in order to dodge the problem, and seemingly at the same time improve the application overall by giving each section a permalink and making the back button work as users navigate through.

**Comment by rodbs** [comment]
I think a good solution for me would be to merge all sections and render them together in a full page (one hydration) but later being able to access  the DOM nodes to hide/unhide each section using `display` css  

Is this feasible?. I've been reading about ref, forwardRef to access DOM elements.
https://stackoverflow.com/questions/38093760/how-to-access-a-dom-element-in-react-what-is-the-equilvalent-of-document-getele

Is it possible to use useRef when combining with hydration? is there a way to pass ref on hydrating on the client?  Or is there a better approach to hide/unhide an already rendered content?
Thanks!

**Comment by jescalan** [comment]
I don't think this would work out the way you want. mdx-remote content is rendered to string and hydrated as part of an entirely separate process from the rest of the react tree, since it's loaded from a string as data, which is the core value proposition of this library.

I think in this case you'd want to just select an id and show and hide it manually. It's not the most react-y way to do things, but I don't think that's the end of the world. If it is, you could always follow my suggestion in the previous comment, which would allow you to deal with everything the way that works best with nextjs convention.

**Comment by rodbs** [comment]
Ok, I think I understand what you mean.
Some final questions:.

1. If I go for your first suggestion of having each section in its own page with its own permalink and buttons to navigate, doesn't it penalize SEO? My idea was to have everything on a same page for SEO purposes.

2. If I decide to render it as single page and select id's manually to hide/unhide, how can I do it? How can I access the DOM manually to hide/unhide sections. I've been reading about it  and they talk about using ref's ... but this is for React Components, not for a string hydrated, right?

3. What about this `react-html-parser` ?
https://medium.com/@vienpham2019/react-html-parser-7bf4286b38f9
Is it feasible to convert the HTML generated by hydrate back to React components? What is your opinion about this solution?

**Comment by jescalan** [comment]
1. No, it does not penalize SEO
2. You can just select the IDs using normal javascript - `document.querySelector` etc.
3. I would not recommend parsing html, there's no way this is worth the cost

**Comment by rodbs** [comment]
I cannot make `document.querySelector`  work  in React.  I manage to access the element, but it doesn't modify the DOM (And I haven't found examples , but all recommend to use refs.)

```
const hide = (elem) => {
    elem.style.display = 'none';
  };
hide(document.getElementById('p2'));
```

Do you know about any examples where modifying the DOM in React using  `document.querySelector` or `document.getElementById`? Is it even possible?
Thanks.

**Comment by jescalan** [comment]
It's definitely possible, but generally not recommended since react does its own DOM management. I do not think this is the direction you want to take things though, overall. I would chase one of the alternate solutions that flows in the natural way that react and nextjs are intended to be used

**Comment by rodbs** [comment]
ok,  understood!

**Comment by jescalan** [comment]
Going to close this one but please feel free to continue the discussion!
